service/src/java/org/apache/hive/service/auth/jwt/JWTValidator.java

```diff
@@ -0,0 +1,110 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hive.service.auth.jwt;
+
+import com.google.common.base.Preconditions;
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSObject;
+import com.nimbusds.jose.JWSVerifier;
+import com.nimbusds.jose.KeyTypeException;
+import com.nimbusds.jose.crypto.ECDSAVerifier;
+import com.nimbusds.jose.crypto.MACVerifier;
+import com.nimbusds.jose.crypto.RSASSAVerifier;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.RSAKey;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import org.apache.hadoop.hive.conf.HiveConf;
+
+import javax.crypto.SecretKey;
+import java.io.IOException;
+import java.security.interfaces.ECPublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+import java.util.Date;
+import java.util.List;
+
+public class JWTValidator {
+
+  private final URLBasedJWKSProvider jwksProvider;
+
+  public JWTValidator(HiveConf conf) throws IOException, ParseException {
+    this.jwksProvider = new URLBasedJWKSProvider(conf);
+  }
+
+  public String validateJWTAndExtractUser(String signedJwt) throws Exception {
+    final SignedJWT parsedJwt = SignedJWT.parse(signedJwt);
+    List<JWK> matchedJWKS = jwksProvider.getJWKs(parsedJwt.getHeader());
+
+    // verify signature
+    boolean signatureVerificationSuccessful = false;
+    for (JWK matchedJWK : matchedJWKS) {
+      JWSVerifier verifier = getVerifier(parsedJwt, matchedJWK);
+      if (parsedJwt.verify(verifier)) {
+        signatureVerificationSuccessful = true;
+        break;
+      }
+    }
+    Preconditions.checkState(signatureVerificationSuccessful, "Unable to verify incoming JWT Signature");
+
+    // verify claims
+    JWTClaimsSet claimsSet = parsedJwt.getJWTClaimsSet();
+    Date expirationTime = claimsSet.getExpirationTime();
+    if (expirationTime != null) {
+      Date now = new Date();
+      Preconditions.checkState(now.before(expirationTime), "JWT has been expired");
+    }
+
+    // We assume the subject of claims is the query user
+    return claimsSet.getSubject();
+  }
+
+
+
+  // TODO Same logic as DefaultJWSVerifierFactory#createJWSVerifier -- This can be written more elegantly, check AGAIN
+  private static JWSVerifier getVerifier(JWSObject parsedJWT, JWK key) throws JOSEException {
```

Comment:
I rewrote the logic of `getVerifier` because I found some bugs of incompatible class casting. We need this function because we need to pass correct arguments into `DefaultJWSVerifierFatory.createJWSVerifier`.